### PR TITLE
Added referenced counted shared pointer class

### PR DESCRIPTION
**Pull request by marcuschangarm**
Class follows the shared_ptr API from C++11. Reference counts are updated when SharedPointer objects are passed around by value. 

Usage: 

```
class A;

SharedPointer<A> objectFactory()
{
    SharedPointer<A> result;

    result = SharedPointer<A>(new A(...));

    return result;
}

...

SharedPointer<A> object1;
SharedPointer<A> object2;

// assignment
object1 = objectFactory();
object2 = object1;

// invoke method in object pointed to
object1->doStuffWithObject();

// invoke method in container
object1.use_count();

```

### CODE REVIEW

**Comment by marcuschangarm** [comment]
@bogdanm I've added the changes you mentioned last time we spoke about this.

File: mbed-util/SharedPointer.h [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+        SHAREDPOINTER_DEBUG("SP&: %p = %p [%p: %p = %lu]\r\n", this, &source, pointer, counter, *counter);
+    }
+
+    /*
+        Assignment operator.
+
+        Cleanup previous reference and assign new pointer and counter.
+    */
+    SharedPointer operator=(const SharedPointer source)
```
Comment (rgrover) [inline mbed-util/SharedPointer.h]: should take in a const reference as an argument

Comment (marcuschangarm) [inline mbed-util/SharedPointer.h]: Which functions are you referring to?

Comment (rgrover) [inline mbed-util/SharedPointer.h]: operator= should take a const reference as an argument. else you'll be creating temporary objects on the side.

Comment (marcuschangarm) [inline mbed-util/SharedPointer.h]: Ah, I see. Thanks! Changed to reference.

**Comment by marcuschangarm** [comment]
see https://github.com/ARMmbed/core-util/pull/30

**Comment by hugovincent** [comment]
Sorry if you've already discussed this but if it follows the `std::shared_ptr` API why not call it `mbed:: shared_ptr` or something?

**Comment by marcuschangarm** [comment]
I wanted to make it clear it is not the same thing. The constructor is different for example. 

And the name resembles FunctionPointer.

**Comment by hugovincent** [comment]
Ah ok, similar but not the same API then.

**Pull request closed without merging**